### PR TITLE
rgw/kafka: do not destroy the connection on errors

### PR DESCRIPTION
**Pull request by yuvalif**
fixes: https://tracker.ceph.com/issues/66017

as well as other simplifications:
* do not store temporary configuration in the connection object. just use as a local variable
* do not create a connection without a producer

other improvements:
* copy to a local list before publishing
* convert internal error codes to errno

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
teuthology passing: https://pulpito.ceph.com/yuvalif-2024-04-03_04:35:49-rgw:notifications-wip-yuval-63909-distro-default-smithi/

**Comment by cbodley** [comment]
i see that https://github.com/ceph/ceph/pull/55051 closed. should https://tracker.ceph.com/issues/63915 point to this pr now?

**Comment by yuvalif** [comment]
@cbodley there is one bug fix here, which is causing a crash-on-close. should i open a tracker for it and link to this PR?

**Comment by cbodley** [comment]
> @cbodley there is one bug fix here, which is causing a crash-on-close. should i open a tracker for it and link to this PR?

yes, since i assume we need backport to squid at least

File: src/rgw/rgw_kafka.cc [hunk 1]
```diff
@@ -441,18 +443,16 @@ class Manager {
...
-  // (4) TODO reconnect on connection errors
-  // (5) TODO cleanup timedout callbacks
   void run() noexcept {
     while (!stopped) {
 
       // publish all messages in the queue
       auto reply_count = 0U;
-      const auto send_count = messages.consume_all(std::bind(&Manager::publish_internal, this, std::placeholders::_1));
+      std::vector<message_wrapper_t*> local_messages;
+      const auto send_count = messages.consume_all([&local_messages](auto message) {local_messages.push_back(message);});
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:452]: `consume_all` is a non-blocking function and prior to your changes, it was invoking `publish_internal` which was run async in diff thread where the `conn` was used. and while below code had a block where it checks to `conn` to be idle and if yes it deletes the connection.
so there were chances of race condition where before `publish_internal` goes and updates the `conn->time` the `idle` check was executed and conn was destroyed causing the race condition and crash when `publish_internal` tries to access the `conn`
if yes, i just have 1 doubt, `consume_all` will still run async and `local_messages` will not be populated immediately and then `std::for_each` would loop over the `local_messages` which would not have populated and this will result in loss of messages being delivered?
with current changes, you are consuming all the messages and then calling the `publish_internal` in `std::for_each` loop so ensuring the `idle` check will be executed only after the `publish_internal` as been completed for all of the messages and then this would avoid the race condition as destroy and publish_internal would never be executed in parallel?

File: src/rgw/rgw_kafka.cc [hunk 2]
```diff
@@ -201,88 +206,88 @@ using connection_t_ptr = std::unique_ptr<connection_t>;
...
+  rd_kafka_conf_set_opaque(conf.get(), conn);
   // redirect kafka logs to RGW
-  rd_kafka_conf_set_log_cb(conn->temp_conf, log_callback);
+  rd_kafka_conf_set_log_cb(conf.get(), log_callback);
   // define poll callback to allow reconnect
-  rd_kafka_conf_set_error_cb(conn->temp_conf, poll_err_callback);
-  // create the producer
-  if (conn->producer) {
-    ldout(conn->cct, 5) << "Kafka connect: producer already exists. detroying the existing before creating a new one" << dendl;
-    conn->destroy(STATUS_CONF_REPLCACE);
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:283]: another thing i noticed is, we call `destroy` inside the `publish_internal` as well and bunch of other places, but after clearing the connection, we do not clear it from the `connection_list` so chances are we read the old connection again and cause crash
so wherever we explicitly call `conn->destroy` we should call `connection_list->clear(conn)`, so we do not read the stale connection

File: src/rgw/rgw_kafka.cc (hunk 1 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:452]: we were running in the same thread (the kafka manager thread) even before the change.
nothing was running in paralel, and `consume_all` is completly syncronouse, so, i don't think there was a risk for race condition.
the main reason for the change is that before the change, traversing the queue was taking longer (since we called `publish_internal` as we traversed the queue.
this means that it is more likely for the thread that pushes into the queue to find it full.
however, i did not see any visible performence improvments with the change, so i'm also ok with reverting it.

File: src/rgw/rgw_kafka.cc (hunk 2 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: this is the main point of the change - after the change the only place where we call `destroy()` in the dtor of the connections.
so, it is invoked only after the connection is removed from the list.

File: src/rgw/rgw_kafka.cc (hunk 1 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:452]: probably reverting the change would make sense then for `consume_all` ??

File: src/rgw/rgw_kafka.cc (hunk 2 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:283]: but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection.
so i still see 2 issues here

- now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.
- this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash

File: src/rgw/rgw_kafka.cc [hunk 3]
```diff
@@ -377,11 +381,11 @@ class Manager {
       topic = rd_kafka_topic_new(conn->producer, message->topic.c_str(), nullptr);
       if (!topic) {
         const auto err = rd_kafka_last_error();
-        ldout(conn->cct, 1) << "Kafka publish: failed to create topic: " << message->topic << " error: " << status_to_string(err) << dendl;
+        ldout(conn->cct, 1) << "Kafka publish: failed to create topic: " << message->topic << " error: "
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:384]: this whole logic of calling `rd_kafka_topic_new` if topic does not exist doesn't make sense.
what could be the scenario where this could happen.
`publish_internal` will always be called after the connection is created ?

File: src/rgw/rgw_kafka.cc (hunk 2 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:283]: > but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection. so i still see 2 issues here
> 
> * now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.
> * this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash

looking at code more, i am still confused on what could be the reason for `realm-reload` cash, coz for manager shutdown, the destructor does call `runner.join();` so all events should have been done publishing and `conn` should be safe to delete and hence no race condition?

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: > > but the problem now is that earlier, for any errors while sending the notification to topic or creating new topic we would explicitly delete the connection (but not from the list) and now for those error condition we do not delete the connection. so i still see 2 issues here
> > 
> > * now for regular use case, whenever there is any error while producing the notification or all other places where we explicitly destroyed the connection before, will hold onto stale connection? coz now we only delete the connection for `IDLE` scenario or `shutdown/realm upload` scenario.

there was never a reason to destroy the connection on error. the librdkafka code have the right mechanisms to overcome such errors and there is no need to destroy the connection when we encounter them
the idle scenarion is to cleanup unused connection (instead of a more complex ref counting mechanism), and has nothing to do with error handling.

> > * this will still not fix the `realm-reload` crash as if there is active sending of notification going and if there is `realm-reload` it will shutdown manager causing the connections to be destroyed in separate thread again causing the crash
> 
> looking at code more, i am still confused on what could be the reason for `realm-reload` cash, coz for manager shutdown, the destructor does call `runner.join();` so all events should have been done publishing and `conn` should be safe to delete and hence no race condition?

realm reload crash is not handled here. would require a different change.
the reason for the crash is that we invoke a callback that releases a coroutine in the notification manager that does not exist anymore (due to the reload).

this PR is to have simpler code that does not add competing logic to the librdkafka logic.
and it fixes a valgrind error that may indicate crash-on-close (not reload).

File: src/rgw/rgw_kafka.cc (hunk 3 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:384]: from kafka perspective "connection" and "topic" are not the same. one connection may serve multiple topics.
if we try to publish to a topic and it does not exist, publishing would fail.
so, to be on the safe side, we create the topic (the kafka topic) first.

File: src/rgw/rgw_kafka.cc (hunk 2 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:283]: so then this PR is technically not fixing any crashes 
its just more about simplification and  potential fix any crashes where the conn is destroyed but not removed from the list and code ends up using the deleted connection and crash ?

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: yes. this is fixing a valgrind error (which could be a crash-on-close) and simplifying the code

Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:283]: got it, so then how is the ` crash is that we invoke a callback that releases a coroutine in the notification manager ` is getting fixed here ? do you mind explaining a little ?

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: it will be fixed in a different PR (working on it...)

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: it is fixing a valgrind memory error that could resulty in crash-on-shutdown (which is a 'low priorty' crash)

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:283]: reload crash fix is here: https://github.com/ceph/ceph/pull/56979

File: src/rgw/rgw_kafka.cc (hunk 3 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:384]: sorry i think i misinterpreted this, 
so kafka needs to call `rd_kafka_topic_new` before calling `rd_kafka_produce`.
but what is confusing is why is this a vector ?
for a given kafka topic, what are the scenario where we will need multiple `rd_kafka_topic_new` ?
this vector list i weird, the  vector is holding `rd_kafka_topic_t` --> `std::vector<rd_kafka_topic_t*> topics`
and while searching `auto topic_it = std::find(conn->topics.begin(), conn->topics.end(), message->topic);` you are searching for a `string` inside the vector ?

so basically i am asking is
-> why do we need vector ? a given topic will have one on one mapping to `rd_kafka_topic_t`, so while we connection caching(pooling) we create the connection and then we need to call `rd_kafka_topic_new` only once ?

**Comment by kchheda3** [comment]
@yuvalif do we not have a tracker for this
we are hitting this crash all the time, and its brining down all the rgw daemons

**Comment by kchheda3** [comment]
> > @yuvalif do we not have a tracker for this
> > we are hitting this crash all the time, and its brining down all the rgw daemons
> 
> This is crash on shutdown, so unlikely tonbe observed. This is mainly a code cleanup pr

nope, we are able to repro this all the time
if the `rd_kafka_produce` produces a `fatal` error it returns `-1` and then [code](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L412) destroys the connection & `rd_kafka_topic_t`, but does not remove it from `vector list` and then next code uses the deleted `topic` from the vector as it was not removed and crashes.
to ask `rd_kafka_produce` produces a `fatal` error, we are able to do it by just revoking the credential of the kafka user.

**Comment by yuvalif** [comment]
> > > @yuvalif do we not have a tracker for this
> > > we are hitting this crash all the time, and its brining down all the rgw daemons
> > 
> > 
> > This is crash on shutdown, so unlikely tonbe observed. This is mainly a code cleanup pr
> 
> nope, we are able to repro this all the time if the `rd_kafka_produce` produces a `fatal` error it returns `-1` and then [code](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L412) destroys the connection & `rd_kafka_topic_t`, but does not remove it from `vector list` and then next code uses the deleted `topic` from the vector as it was not removed and crashes. to ask `rd_kafka_produce` produces a `fatal` error, we are able to do it by just revoking the credential of the kafka user. was looking at the `rd_kafka_produce` and there are multiple use case when the `rd_kafka_produce` can return fatal error (-1), so its not only about the shutdown

i never saw this issue. but this just make this fix even more critical.
can you please open a tracker, with the above scenario, so we can backport the fix, at least to squid?

**Comment by kchheda3** [comment]
> > > > @yuvalif do we not have a tracker for this
> > > > we are hitting this crash all the time, and its brining down all the rgw daemons
> > > 
> > > 
> > > This is crash on shutdown, so unlikely tonbe observed. This is mainly a code cleanup pr
> > 
> > 
> > nope, we are able to repro this all the time if the `rd_kafka_produce` produces a `fatal` error it returns `-1` and then [code](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L412) destroys the connection & `rd_kafka_topic_t`, but does not remove it from `vector list` and then next code uses the deleted `topic` from the vector as it was not removed and crashes. to ask `rd_kafka_produce` produces a `fatal` error, we are able to do it by just revoking the credential of the kafka user. was looking at the `rd_kafka_produce` and there are multiple use case when the `rd_kafka_produce` can return fatal error (-1), so its not only about the shutdown
> 
> i never saw this issue. but this just make this fix even more critical. can you please open a tracker, with the above scenario, so we can backport the fix, at least to squid?

yeah 100%, we need to backport it to squid for sure
i will open the tracker and i was planning to ping you for this issue only
i had few questions on the maintaining that vector which i already commented

File: src/rgw/rgw_kafka.cc (hunk 3 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:384]: an RGW topic contains 2 main things: the endpoint and the topic name.
the kafka manager holds a list of connection (=endpoints) each has a vector of kafka topics.
lets assume we have the following scenarion:
RGW topic1: kafka://host1:9092, mytopic1
RGW topic2: kafka://host1:9092, mytopic2

so, in the kafka manager, we will have one connection (to host1:9092) with a vector of 2 topics: topic1, topic2
when we send the notifications, once we created the 2 topics on the broker, we will see that they are in the vector, and we won't create them again.
note, that when the RGW topics are created, the kafka manager has no information about that. only when it sends the notifications it opens the connection and creates the topics

Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:384]: got it, its just the name of the topic that is passed during the `CreateTopic`
but again why maintain the vector ? coz they both are pointing to same connection? can't both topics use the same connection ?

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:384]: inside the kafka manager we do not masintain a list of RGW topics.
we maintain a list of kafka endpoints, and each endpoint is holding a vector of topics.
this means that if we have multipel RGW topics that share a connection, they will become a single connection with a vector of topic names in the kafka manager

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/rgw_kafka.cc (hunk 3 above)
Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:384]: Yep I got that logic, Kafka manager topic can hold multiple rgw topics , what I was asking is why can't multiple rgw topics map to single Kafka connection ?
Is it necessary to maintain this vector ?

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:384]: The kafka manager does not know anything about rgw topics.
It hold a list of connection (=brokers). Each holding a list of kafka topics pn that broker.

Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:384]: > The kafka manager does not know anything about rgw topics. It hold a list of connection (=brokers). Each holding a list of kafka topics pn that broker.

@yuvalif we are repeating the same thing,
kafka manager maintains a map based on endpoint/broker.
the 2 rgw topics point to same endpoint/broker, so both rgw topics can technically use the same connection if we do not create the vector? is there any reason we map 2 rgw topics to single kafka manager connection? can't we use same

File: src/rgw/rgw_kafka.cc [hunk 4]
```diff
@@ -79,40 +118,29 @@ struct connection_t {
...
-        rd_kafka_conf_destroy(temp_conf);
-        return;
-    }
-    if (!is_ok()) {
+  void destroy() {
+    if (!producer) {
       // no producer, nothing to destroy
       return;
     }
-    // wait for all remaining acks/nacks
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:93]: you still need to call `rd_kafka_flush` to flush out the remaining callbacks

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:93]: I'm clearing the remaining callbacks (with an error message) in line 133. so there is no issue with clearing the callbacks.
if the broker is ok and shutdown process is called due to a reload (since  https://github.com/ceph/ceph/pull/56979 is already merged) `rd_kafka_flush()` will invoke the callbacks with success.
OTOH, if the broker is not ok, it may delay the reload process in another 5 seconds, so I'm not sure it is a good idea.

Comment (kchheda3) [reply src/rgw/rgw_kafka.cc:93]: > if the broker is ok and shutdown process is called d

nope, i was talking in context of shutdown, thats the only other route from where the `destroy` will now be invoked other than `IDLE`, so in shutdown case we just wait additional 5 seconds for callbacks to process, should be ok ?

File: src/rgw/rgw_kafka.cc [hunk 5]
```diff
@@ -166,7 +171,7 @@ void message_callback(rd_kafka_t* rk, const rd_kafka_message_t* rkmessage, void*
   if (tag_it != callbacks_end) {
       ldout(conn->cct, 20) << "Kafka run: n/ack received, invoking callback with tag=" << 
           *tag << " and result=" << rd_kafka_err2str(result) << dendl;
-      tag_it->cb(result);
+      tag_it->cb(-rd_kafka_err2errno(result));
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:174]: @yuvalif we did testing with these changes, and i feel we should not map the kafka error's to linux errors.
i do not see a reason on why we want to to convert these error's to linux error's, coz currently for any reason if there is any config issue with kafka, you get to know the error code on what the error is in the callback return code.
with error transition to linux based error, the only way to know the error is by turning on the debug level 20 which then prints the kafka error string `rd_kafka_err2str`.
i do not see any value by converting them linux error's.

Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:174]: the reason for conversing to linux errors is that the calling code is not kafka specific.
e.g. we can reply them back to the client ion case of non-persistent notifications.

agree that we lose valuable info here (especially in case of persistent notification). 
will fix that, so we log the actual kafka errors

File: src/rgw/rgw_kafka.cc (hunk 4 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:93]: i'll try to change and test

File: src/rgw/rgw_kafka.cc (hunk 1 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc]: done

File: src/rgw/rgw_kafka.cc (hunk 4 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:93]: added back flushing.

File: src/rgw/rgw_kafka.cc (hunk 5 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:184]: making sure that the kafka errors are printed at log level 1

File: src/rgw/rgw_kafka.cc [hunk 6]
```diff
@@ -79,40 +118,31 @@ struct connection_t {
...
     producer = nullptr;
     // fire all remaining callbacks (if not fired by rd_kafka_flush)
     std::for_each(callbacks.begin(), callbacks.end(), [this](auto& cb_tag) {
-        cb_tag.cb(status);
-        ldout(cct, 20) << "Kafka destroy: invoking callback with tag="
+        ldout(cct, 1) << "Kafka destroy: invoking callback with tag: "
                        << cb_tag.tag << " for: " << broker
-                       << " with status: " << status << dendl;
+                       << " with status: " << status_to_string(status) << dendl;
+        cb_tag.cb(status_to_errno(status));
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:139]: status default value = 0.
so this will evict the entries from the queue,
probably explicitly send `DESTROY` or some sensible message here ?

File: src/rgw/rgw_kafka.cc [hunk 7]
```diff
@@ -412,12 +426,10 @@ class Manager {
     if (rc == -1) {
       const auto err = rd_kafka_last_error();
       ldout(conn->cct, 10) << "Kafka publish: failed to produce: " << rd_kafka_err2str(err) << dendl;
-      // TODO: dont error on full queue, and don't destroy connection, retry instead
       // immediatly invoke callback on error if needed
       if (message->cb) {
-        message->cb(err);
+        message->cb(-rd_kafka_err2errno(err));
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:431]: can we also make the log_level =1 for line no. 428, since its an error and would like to know the kafka error

File: src/rgw/rgw_kafka.cc (hunk 6 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:139]: right. destroy() is called either upon idleness (where state will be set accordingly).
or on shutdown (where state reflect last error or just zero).
will set state on shutdown

File: src/rgw/rgw_kafka.cc (hunk 4 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:93]: but reduced to 500ms (compromise between not blocking reloads to letting the kafka broker reply correctly)

File: src/rgw/rgw_kafka.cc (hunk 6 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:139]: changed status of all connections i nthe manager's dtor

**Comment by yuvalif** [comment]
teuthology passing (kafka tests): https://pulpito.ceph.com/yuvalif-2024-05-23_10:33:57-rgw:notifications-wip-yuval-kafka-cleanup-distro-default-smithi/

**Comment by yuvalif** [comment]
jenkins test make check

**Comment by yuvalif** [comment]
jenkins test docs

**Comment by yuvalif** [comment]
jenkins render docs

**Comment by yuvalif** [comment]
jenkins test windows

File: src/rgw/rgw_kafka.cc [hunk 8]
```diff
@@ -121,41 +151,26 @@ struct connection_t {
...
 
   const auto conn = reinterpret_cast<connection_t*>(opaque);
   const auto result = rkmessage->err;
 
+  if (rkmessage->err == 0) {
+      ldout(conn->cct, 20) << "Kafka run: ack received with result=" << 
+        rd_kafka_err2str(result) << dendl;
+  } else {
+      ldout(conn->cct, 1) << "Kafka run: nack received with result=" << 
+        rd_kafka_err2str(result) << dendl;
```
Comment (kchheda3) [inline src/rgw/rgw_kafka.cc:169]: nit, if you could add the broker and topic name here and also for error after calling `rd_kafka_produce` at line 428
"topic: " << rd_kafka_name(rk) << " broker: " << conn->broker << dendl

**Review by kchheda3 — APPROVED** [review]
Thanks @yuvalif

File: src/rgw/rgw_kafka.cc (hunk 8 above)
Comment (yuvalif) [reply src/rgw/rgw_kafka.cc:169]: will do a round of debug log fixes (in a different PR), and add that there.
* use DoutPrefixProvider
* unify what log messages contain (broker+topic)
